ground/src/main/java/com/google/android/ground/ui/map/MapController.kt

```diff
@@ -43,50 +50,51 @@ constructor(
   private val locationOfInterestRepository: LocationOfInterestRepository,
   private val surveyRepository: SurveyRepository,
   private val mapStateRepository: MapStateRepository,
-  @DefaultDispatcher private val defaultDispatcher: CoroutineDispatcher,
+  @ApplicationScope private val externalScope: CoroutineScope
 ) {
 
-  private val cameraUpdatesSubject: @Hot Subject<CameraPosition> = PublishSubject.create()
+  private val _cameraUpdates = MutableStateFlow<CameraPosition?>(null)
 
   /** Emits a combined stream of camera update requests. */
-  fun getCameraUpdates(): Flowable<CameraPosition> =
-    cameraUpdatesSubject
-      .toFlowable(BackpressureStrategy.LATEST)
-      .mergeWith(getCameraUpdatesFromLocationChanges())
-      .mergeWith(getCameraUpdatedFromSurveyChanges())
+  fun getCameraUpdates(): SharedFlow<CameraPosition> =
+    merge(
+        _cameraUpdates.filterNotNull(),
+        getCameraUpdatesFromLocationChanges(),
+        getCameraUpdatedFromSurveyChanges(),
+      )
+      .shareIn(externalScope, replay = 1, started = SharingStarted.Eagerly)
 
   /** Emits a stream of camera update requests due to location changes. */
-  private fun getCameraUpdatesFromLocationChanges(): Flowable<CameraPosition> {
-    val locationUpdates = locationManager.locationUpdates.asFlowable().map { it.toCoordinates() }
+  private fun getCameraUpdatesFromLocationChanges(): Flow<CameraPosition> {
+    val locationUpdates = locationManager.locationUpdates.map { it.toCoordinates() }
     // The first update pans and zooms the camera to the appropriate zoom level;
     // subsequent ones only pan the map.
     return locationUpdates
       .take(1)
       .map { CameraPosition(it, DEFAULT_LOI_ZOOM_LEVEL) }
-      .concatWith(locationUpdates.map { CameraPosition(it) }.skip(1))
+      .onCompletion { emitAll(locationUpdates.map { CameraPosition(it) }.drop(1)) }
```

Comment:
Can we simplify this flow by using `withIndex`? In that case we'd update the zoom level only when the index is 0.

Reply:
Thanks for the tip! I didn't know of this method. Makes it so much better.